### DISCUSSION TITLE
build problems CODESYS

### DISCUSSION THREAD

**Discussion by Nasimovy** [MichielVanwelsenaere/HomeAutomation.CoDeSys3#96] [Q&A]
the RTC_2 library from Basic is giving me troubles

![image](https://user-images.githubusercontent.com/20001575/136688108-4c7ead22-40e8-4110-b471-7fa56677ec74.png)

has anyone else encountered this problem before ?

**Comment by MichielVanwelsenaere** [comment]
Hi,

the problem you are describing is due to 'LDT' being introduced as a datatype in newer Codesys versions. Unfortunately there's a variable named 'LDT' in the OSCAT libraries causing troubles.

This is described further in #76 , as described, updating the OSCAT libs to the latest version should fix the problem.
Let me know if you run into any other problems.

**Reply by Nasimovy** [reply]
i downloaded and installed all the updates but codesys doesnt let me choose them
![image](https://user-images.githubusercontent.com/20001575/136998675-febdd035-79f7-427d-81f5-bbeb549bad10.png)

oscat network im not able to select the newest version, and there are other basics that are not the newest version

**Reply by MichielVanwelsenaere** [reply]
ahn right, I forgot that that dependency was nested in the MQTT library as well. Stefan has created a newer version of the MQTT library with improved dependency handling but I haven't found the time to incorporate it here.

you have two options:
- you can attempt yourself to update the MQTT library from Stefan and fix the dependency problem
- downgrade to CodeSys V3.5 SP15 (the 'LDT' datatype got introduced from SP16)

**Reply by MichielVanwelsenaere** [reply]
if you are using é!Cockpit you could also try setting the compiler version to the number below:
![image](https://user-images.githubusercontent.com/6965657/137178623-3fe563ee-9aba-40d7-93f8-02a833ad0cf3.png)

**Comment by Nasimovy** [comment]
i did the downgrade but now i cant login to the pfc to add the io modules 
![image](https://user-images.githubusercontent.com/20001575/137345442-926d1e6f-3e7e-4252-a9ab-60da7817ed54.png)

**Reply by MichielVanwelsenaere** [reply]
are you sure your device is running the same runtime version as the compiler version you are currently running?

**Reply by Nasimovy** [reply]
Not sure, I'm not even sure were to check that

**Reply by MichielVanwelsenaere** [reply]
Does it show up with wago IO check? And if so, with what Runtime and what version?

I'm currently in the middle of a construction project, my PLC's are being wired and network rebuild, you'll have to feed me information as I don't have a device available for testing purposes right now.
